src/osmet.rs

```diff
@@ -0,0 +1,809 @@
+            blksize
+        );
+    }
+
+    println!("Total bytes skipped: {}", total_bytes_punched);
+    println!("Total bytes written: {}", total_bytes_written);
+    println!("Total bytes written (compressed): {}", xz_tmpf.total_out());
+
+    let mut tmpf = xz_tmpf.finish()?;
+    tmpf.seek(SeekFrom::Start(0))?;
+
+    Ok((tmpf, blksize.get()))
+}
+
+fn write_punched_image(
+    dev: &mut File,
+    w: &mut impl Write,
+    partitions: &[OsmetPartition],
+) -> Result<u64> {
+    let mut buf = [0u8; 8192];
+
+    let mut cursor: u64 = 0;
+    let mut total_bytes_punched: u64 = 0;
+    for partition in partitions {
+        // first copy everything up to the start of the partition
+        assert!(partition.start_offset >= cursor);
+        cursor += copy_exactly_n(dev, w, partition.start_offset - cursor, &mut buf)?;
+        total_bytes_punched += write_punched_image_partition(dev, w, partition, &mut buf)?;
+    }
+
+    // and finally write out the remainder of the disk
+    io::copy(dev, w)?;
+
+    Ok(total_bytes_punched)
+}
+
+fn write_punched_image_partition(
+    dev: &mut File,
+    w: &mut impl Write,
+    partition: &OsmetPartition,
+    buf: &mut [u8],
+) -> Result<u64> {
+    let mut total_bytes_punched = 0;
+
+    // and this is where the real fun begins!
+    let mut cursor = partition.start_offset;
+    for mapping in partition.mappings.iter() {
+        // make offset relative to start of disk, not partition
+        let extent_start = mapping.extent.physical + partition.start_offset;
+        assert!(extent_start >= cursor);
+        if cursor < extent_start {
+            cursor += copy_exactly_n(dev, w, extent_start - cursor, buf)
+                .chain_err(|| "while writing in between extents")?;
+        }
+
+        // this is the crucial space-saving step; we skip over the extent we have a mapping for
+        dev.seek(SeekFrom::Current(mapping.extent.length.try_into().unwrap()))
+            .chain_err(|| format!("while skipping extent: {:?}", mapping.extent))?;
+        total_bytes_punched += mapping.extent.length;
+        cursor += mapping.extent.length;
+    }
+
+    assert!(cursor <= partition.end_offset);
+
+    // and now just transfer the rest of the the partition and disk
+    copy_exactly_n(dev, w, partition.end_offset - cursor, buf)
+        .chain_err(|| "while writing remainder")?;
+
+    Ok(total_bytes_punched)
+}
+
+fn write_osmet_file(path: &Path, osmet: Osmet, mut punched_image: File) -> Result<()> {
+    // would be nice to opportunistically do open(O_TMPFILE) then linkat here, but the tempfile API
+    // doesn't provide that API: https://github.com/Stebalien/tempfile/pull/31
+    let mut f = BufWriter::new(NamedTempFile::new_in(path.parent().unwrap())?);
+
+    let osmet_file = OsmetFile {
+        magic: OSMET_FILE_HEADER_MAGIC,
+        major: 0,
+        minor: 1,
+        osmet,
+    };
+
+    bincode::serialize_into(&mut f, &osmet_file).chain_err(|| "failed to serialize osmet file")?;
```

Comment:
I'm a bit worried about using `bincode` here for data at rest, as I fear it has no interop story outside of Rust. Can we pick somebody a bit more common?

Reply:
Hmm, do you have any recommendations? I really like `bincode`'s dead-simple approach and am not sure if interop outside of Rust should be a concern given how specific to this code the binary format is. (It's not like a generic JSON file which would be useful on its own.)

Reply:
I somehow see this as an interop wrapper between the containerized producing side (cosa/mantle) and the on-host consuming side (coreos-installer). Clearly not the same concerns as public metadata, but still something that benefits from a stable and easy format for both sides. I had protobuf in mind, even if I'm not a great fan of it.

Reply:
Do the discussions in https://github.com/coreos/coreos-installer/pull/187#discussion_r392888346 clear this up?

Reply:
The comment at https://github.com/coreos/coreos-installer/pull/187#issuecomment-607445367 did. We are taking all precautions so that (so far) this is not in any way a public API.